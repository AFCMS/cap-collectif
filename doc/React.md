### How to write localizable components

Just import the appropriate [component](https://github.com/yahoo/react-intl/wiki#the-react-intl-module) from `react-intl`

- For localizable text use
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage).

- For date and time:
[`<FormattedDate>`](https://github.com/yahoo/react-intl/wiki/Components#formatteddate)
[`<FormattedTime>`](https://github.com/yahoo/react-intl/wiki/Components#formattedtime)
[`<FormattedRelative>`](https://github.com/yahoo/react-intl/wiki/Components#formattedrelative)

- For numbers and currencies:
[`<FormattedNumber>`](https://github.com/yahoo/react-intl/wiki/Components#formattednumber)
[`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural)

- If possible, do not use `<FormattedHTMLMessage>`, see how to use *Rich Text Formatting* with
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

- When you need an imperative formatting API, use the [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injectintl) High-Order Component.
