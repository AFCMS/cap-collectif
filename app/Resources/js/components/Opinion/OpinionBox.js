import React, { PropTypes } from 'react';
import { IntlMixin } from 'react-intl';
import OpinionPreview from './OpinionPreview';
import OpinionAnswer from './OpinionAnswer';
import OpinionButtons from './OpinionButtons';
import OpinionAppendices from './OpinionAppendices';
import OpinionBody from './OpinionBody';
import OpinionVotesBox from './Votes/OpinionVotesBox';

const OpinionBox = React.createClass({
  propTypes: {
    opinion: PropTypes.object.isRequired,
    rankingThreshold: PropTypes.number,
    opinionTerm: PropTypes.number,
  },
  mixins: [IntlMixin],

  getMaxVotesValue() {
    return this.getOpinionType().votesThreshold;
  },

  getOpinionType() {
    return this.isVersion() ? this.props.opinion.parent.type : this.props.opinion.type;
  },

  getBoxLabel() {
    return this.isVersion() ? 'opinion.header.version'
      : this.props.opinionTerm === 0
        ? 'opinion.header.opinion'
        : 'opinion.header.article'
    ;
  },

  isVersion() {
    return this.props.opinion && this.props.opinion.parent;
  },

  render() {
    const opinion = this.props.opinion;
    const color = this.getOpinionType().color;
    const backLink = this.isVersion() ? opinion.parent._links.show : opinion._links.type;
    const backTitle = this.isVersion() ? opinion.parent.title : this.getOpinionType().title;
    const headerTitle = this.getBoxLabel();

    const colorClass = 'opinion opinion--' + color + ' opinion--current';
    return (
      <div className="block block--bordered opinion__details">
        <div className={colorClass}>
          <div className="opinion__header opinion__header--centered">
            <a className="pull-left btn btn-default opinion__header__back" href={backLink}>
              <i className="cap cap-arrow-1-1"></i>
              <span className="hidden-xs hidden-sm"> {backTitle}</span>
            </a>
            <h2 className="h4 opinion__header__title">{this.getIntlMessage(headerTitle)}</h2>
          </div>
          <OpinionPreview rankingThreshold={this.props.rankingThreshold} opinionTerm={this.props.opinionTerm} opinion={opinion} link={false} />
        </div>
        <OpinionAppendices opinion={opinion} />
        <div className="opinion__description">
          <p className="h4" style={{ marginTop: '0' }}>{opinion.title}</p>
          <OpinionBody opinion={opinion} />
          <div className="opinion__buttons" style={{ marginTop: '15px', marginBottom: '15px' }} aria-label={this.getIntlMessage('vote.form')}>
            <OpinionButtons opinion={opinion} />
          </div>
          <OpinionVotesBox opinion={opinion} />
        </div>
        <OpinionAnswer answer={opinion.answer} />
      </div>
    );
  },

});

export default OpinionBox;
