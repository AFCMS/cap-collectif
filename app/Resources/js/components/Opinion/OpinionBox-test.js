/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';

import { OpinionBox } from './OpinionBox';

describe('<OpinionBox />', () => {
  const props = {
    opinion: {
      title: 'title',
      answer: {},
      type: { title: 'opinionType', color: 'info' },
    },
    rankingThreshold: 0,
    opinionTerm: 0,
  };

  it('should render correctly', () => {
    const wrapper = shallow(<OpinionBox {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
