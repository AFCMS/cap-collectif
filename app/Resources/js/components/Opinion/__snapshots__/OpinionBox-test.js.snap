// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OpinionBox /> should render correctly 1`] = `
<div
  className="block block--bordered opinion__details"
>
  <div
    className="opinion opinion--info opinion--current"
  >
    <div
      className="opinion__header opinion__header--centered"
      style={
        Object {
          "height": "auto",
        }
      }
    >
      <a
        className="pull-left btn btn-default opinion__header__back"
        href="#"
      >
        <i
          className="cap cap-arrow-1-1"
        />
        <span
          className="hidden-xs hidden-sm"
        >
           
          <FormattedMessage
            id="opinion.header.back"
            values={Object {}}
          />
        </span>
      </a>
      <div
        className="opinion__header__title"
      />
      <h2
        className="h4 opinion__header__title"
      >
        <FormattedMessage
          id="opinion.header.opinion"
          values={Object {}}
        />
        <p
          className="small excerpt"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          sectionTitle
        </p>
      </h2>
    </div>
    <ListGroup
      bsClass="list-group"
      className="list-group-custom mb-0"
    >
      <ListGroupItem
        bsClass="list-group-item"
        className="list-group-item__opinion no-border"
        listItem={false}
      >
        <div
          className="left-block"
        >
          <ForwardRef(Relay(OpinionPreview))
            link={false}
            opinion={
              Object {
                "$fragmentRefs": null,
                "$refType": null,
                "__typename": "Opinion",
                "section": Object {
                  "color": "info",
                  "title": "sectionTitle",
                  "url": "#",
                },
                "title": "title",
              }
            }
            opinionTerm={0}
            rankingThreshold={0}
          />
        </div>
      </ListGroupItem>
    </ListGroup>
  </div>
  <ForwardRef(Relay(OpinionAppendices))
    opinion={
      Object {
        "$fragmentRefs": null,
        "$refType": null,
        "__typename": "Opinion",
        "section": Object {
          "color": "info",
          "title": "sectionTitle",
          "url": "#",
        },
        "title": "title",
      }
    }
  />
  <div
    className="opinion__description"
  >
    <ForwardRef(Relay(OpinionBody))
      opinion={
        Object {
          "$fragmentRefs": null,
          "$refType": null,
          "__typename": "Opinion",
          "section": Object {
            "color": "info",
            "title": "sectionTitle",
            "url": "#",
          },
          "title": "title",
        }
      }
    />
    <ForwardRef(Relay(OpinionVotesBox))
      opinion={
        Object {
          "$fragmentRefs": null,
          "$refType": null,
          "__typename": "Opinion",
          "section": Object {
            "color": "info",
            "title": "sectionTitle",
            "url": "#",
          },
          "title": "title",
        }
      }
    />
    <div
      aria-label={
        <FormattedMessage
          id="vote.form"
          values={Object {}}
        />
      }
      className="opinion__buttons"
      style={
        Object {
          "marginBottom": "15px",
          "marginTop": "15px",
        }
      }
    >
      <ForwardRef(Relay(Connect(OpinionButtons)))
        opinion={
          Object {
            "$fragmentRefs": null,
            "$refType": null,
            "__typename": "Opinion",
            "section": Object {
              "color": "info",
              "title": "sectionTitle",
              "url": "#",
            },
            "title": "title",
          }
        }
      />
    </div>
  </div>
  <ForwardRef(Relay(OpinionAnswer))
    opinion={
      Object {
        "$fragmentRefs": null,
        "$refType": null,
        "__typename": "Opinion",
        "section": Object {
          "color": "info",
          "title": "sectionTitle",
          "url": "#",
        },
        "title": "title",
      }
    }
  />
</div>
`;
